stock_inventory_revaluation/models/stock_inventory_revaluation.py

```diff
@@ -0,0 +1,273 @@
+        )
+
+    name = fields.Char(
+        "Name", default=lambda self: _("New"), copy=False, readonly=True, tracking=True
+    )
+    add_from_stock_move = fields.Many2one("stock.move", "Add From Stock Move")
+    date = fields.Date(
+        "Date",
+        default=fields.Date.context_today,
+        copy=False,
+        required=True,
+        states={"done": [("readonly", True)]},
+        tracking=True,
+    )
+    stock_inventory_revaluation_line_ids = fields.One2many(
+        "stock.inventory.revaluation.line",
+        "stock_inventory_revaluation_id",
+        "Revaluation Lines",
+        copy=True,
+        states={"done": [("readonly", True)]},
+    )
+    state = fields.Selection(
+        [("draft", "Draft"), ("done", "Posted"), ("cancel", "Cancelled")],
+        "State",
+        default="draft",
+        copy=False,
+        readonly=True,
+        tracking=True,
+    )
+    company_id = fields.Many2one(
+        "res.company", string="Company", related="vendor_bill_id.company_id"
+    )
+
+    vendor_bill_id = fields.Many2one(
+        "account.move",
+        string="Vendor Bill",
+        help="Vendor Bill that triggers the revaluation",
+        copy=False,
+        domain=[("move_type", "=", "in_invoice")],
+    )
+    account_move_id = fields.Many2one(
+        "account.move", string="Revaluation entry", readonly=True, copy=False
+    )
+    currency_id = fields.Many2one("res.currency", related="company_id.currency_id")
+
+    @api.onchange("add_from_stock_move")
+    def onchange_add_from_stock_move(self):
+        if self.add_from_stock_move:
+            val_line_values = self.get_valuation_lines(self.add_from_stock_move)
+            for vals in val_line_values:
+                self.env["stock.inventory.revaluation.line"].create(vals)
+            self.add_from_stock_move = False
+
+    @api.model
+    def create(self, vals):
+        if vals.get("name", _("New")) == _("New"):
+            vals["name"] = self.env["ir.sequence"].next_by_code(
+                "stock.inventory.revaluation"
+            )
+        return super().create(vals)
+
+    def unlink(self):
+        self.button_cancel()
+        return super().unlink()
+
+    def button_cancel(self):
+        if any(revaluation.state == "done" for revaluation in self):
+            raise UserError(
+                _(
+                    "Validated revaluation cannot be cancelled, but you could revaluation again"
+                )
+            )
+        return self.write({"state": "cancel"})
+
+    def button_validate(self):
+        for revaluation in self:
+            revaluation = revaluation.with_company(revaluation.company_id)
+            move = self.env["account.move"]
+            move_vals = {
+                "journal_id": revaluation.vendor_bill_id.journal_id.id,
+                "date": revaluation.date,
+                "ref": revaluation.name,
+                "line_ids": [],
+                "move_type": "entry",
+            }
```

Comment:
can be separate into new function for purpose inherit if necessary on any custom module?

Reply:
Sure! I just did that :)